### PR TITLE
[SPARK-15056] [SQL] Parse Unsupported Sampling Syntax and Issue Better Exceptions

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

Compared with the current Spark parser, there are two extra syntax are supported in Hive for sampling
- In `On` clauses, `rand()` is used for indicating sampling on the entire row instead of an individual column. For example, 
  
  ``` SQL
  SELECT * FROM source TABLESAMPLE(BUCKET 3 OUT OF 32 ON rand()) s;
  ```
- Users can specify the total length to be read. For example, 
  
  ``` SQL
  SELECT * FROM source TABLESAMPLE(100M) s;
  ```

Below is the link for references:
   https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Sampling

This PR is to parse and capture these two extra syntax, and issue a better error message.
#### How was this patch tested?

Added test cases to verify the thrown exceptions

### CODE REVIEW

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -920,6 +921,10 @@ TINYINT_LITERAL
     : DIGIT+ 'Y'
     ;
 
+BYTELENGTH_LITERAL
+    : DIGIT+ ('b' | 'B' | 'k' | 'K' | 'm' | 'M' | 'g' | 'G')
```
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: The lexer is case insensitive so we don't need to add both lower and upper case characters. Just keep the uppercase ones (for consistency).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -655,8 +655,13 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
         val fraction = ctx.percentage.getText.toDouble
         sample(fraction / 100.0d)
 
+      case SqlBaseParser.BYTELENGTH_LITERAL =>
+        throw new ParseException(
+          "TABLESAMPLE(BUCKET byteLengthLiteral) is not supported", ctx)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: BUCKET? It is a bytetype literal right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -655,8 +655,13 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
...
+
       case SqlBaseParser.BUCKET if ctx.ON != null =>
-        throw new ParseException("TABLESAMPLE(BUCKET x OUT OF y ON id) is not supported", ctx)
+        throw new ParseException(
+          "TABLESAMPLE(BUCKET x OUT OF y ON colname/rand()) is not supported", ctx)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Message is slightly confusing if user is passing in anything other than rand. How about two error messages? One for column and one for function.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Yeah, will do. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: : ) Yeah!

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Thanks! Do it now.

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
@hvanhovell Could you please check if the code changes are appropriate? Thanks!

**Comment by hvanhovell** [comment]
LGTM

**Comment by hvanhovell** [comment]
merging to master/2.0 thanks

**Comment by gatorsmile** [comment]
Thanks!

**Pull request closed without merging**